### PR TITLE
[tune] fix long running release test WIP

### LINKED ISSUE #13953: [release][tune] pytorch pbt failure frozen
### What is the problem?

When running the pytorch pbt release test, the test froze after < 5hrs and has now been frozen for over 5 hrs. This presumably fails the "ensure that it is printing new iterations" requirement. 

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by krfricke**
## Related issue number

Closes #13953

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/utils/mock.py [hunk 1]
```diff
@@ -108,14 +109,22 @@ class FailureInjectorCallback(Callback):
     def __init__(self,
                  config_path="~/ray_bootstrap_config.yaml",
                  probability=0.1,
+                 time_between_checks=0,
```
Comment (krfricke) [inline python/ray/tune/utils/mock.py:112]: Because `on_step_begin` is called much more often with placement groups, we need to introduce some time-based checks here

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 2]
```diff
@@ -647,23 +646,16 @@ def default_resource_request(cls, config):
...
-                    extra_gpu=int(int(use_gpu) * remote_worker_count))
+                bundles = [
+                    # Worker bundles
+                    {
+                        "CPU": num_cpus_per_worker,
+                        "GPU": int(use_gpu)
+                    }
+                ] * num_workers
+
+                return PlacementGroupFactory(bundles, strategy="PACK")
```
Comment (amogkam) [inline python/ray/util/sgd/torch/torch_trainer.py:658]: Isn't the first bundle for the `Trainable` itself, so it should not request any GPUs or `num_cpus_per_worker` CPUs? Is it possible for the first bundle to just be 0 CPUs?

**Review by amogkam — APPROVED** [review]
Lgtm overall, but just left one comment on RaySGD+Tune placement group resources usage.

File: python/ray/util/sgd/torch/torch_trainer.py (hunk 2 above)
Comment (krfricke) [reply python/ray/util/sgd/torch/torch_trainer.py:658]: No, zero-resource bundles raise an error on the ray core side. Previously I added a 1 CPU bundle if local_mode is False, but this would then require more resources than before.

Comment (amogkam) [reply python/ray/util/sgd/torch/torch_trainer.py:658]: I see. Then having the 1 CPU bundle is `use_local` is False is what we want right? Otherwise the Trainable itself will request too many resources, and there will be less bundles than num_workers?

Comment (krfricke) [reply python/ray/util/sgd/torch/torch_trainer.py:658]: Will it though? Previously we didn't request a CPU for the driver process and `num_cpus=0` for `local_mode=False` - e.g. the number of requested resources should be the same here. 

The difference for the `local_mode` is thus just here:

```
        if self.use_local:
            self.worker_group = LocalWorkerGroup(**worker_args)
        else:
            self.worker_group = RemoteWorkerGroup(**worker_args)
 ```
 
 The local trainer of the LocalWorkerGroup will live in the same bundle as the driver, but that shouldn't be a problem as long as the driver doesn't occupy any resource. Or am I missing something here?

Comment (amogkam) [reply python/ray/util/sgd/torch/torch_trainer.py:658]: My understanding is that when using placement groups with Tune, the first bundle is for the Trainable and any subsequent bundles will be used for additional actors that are launched by the trainable. When using RaySGD with Tune, the driver creates the Tune Trainables, and each Tune Trainable instantiates a `TorchTrainer` which creates `num_workers` actors. In the case where `use_local` is True, the first training worker will be on the same process as the Trainable, so the Trainable needs enough resources to launch a worker and the current code works for that case. However when `use_local` is False, all of the workers are remote actors, so the Trainable itself should only request 1 CPU any not any GPUs. So I believe the bundles for the `use_local` False case should be `[{"CPU": 1}] + [{"CPU": cpus_per_worker, "GPU": int(use_gpu)}] * num_workers`

Comment (krfricke) [reply python/ray/util/sgd/torch/torch_trainer.py:666]: Ok so generally this should still work, as the `use_local` just switches between calling the first worker in a `.remote()` call or adding it directly in the driver.

I.e. this code works nicely:

```
import ray
from ray.util import placement_group

ray.init(num_cpus=4)


@ray.remote(num_cpus=4)
def worker_task():
    print(ray.get_resource_ids())
    return 1


@ray.remote
def driver_task():
    print(ray.get(worker_task.remote()))
    return 2


pg = placement_group([{"CPU": 4}])
print(ray.get(driver_task.options(placement_group=pg, num_cpus=0).remote()))
```

However, you are right: In the current placement group implementation the driver process will always occupy the full resources of the first bundle, so there wouldn't be resources left for the remote trainable.

Long story short: Thanks for catching this! I adjusted the code :-)

**Comment by richardliaw** [comment]
Looks like tests are failing unfortunately

**Comment by amogkam** [comment]
@krfricke is moving sgd+tune to placement groups necessary for the release test to work? theres some sgd tests failing right and I think we should separate the placement group stuff to a separate pr so we can cherry pick the release test fix asap

**Comment by krfricke** [comment]
It seems the only thing we need to change is setting `num_cpus=3` (instead of 2) to account for the additional driver CPU (as discussed above).
